xbmc/messaging/ApplicationMessenger.cpp

```diff
@@ -225,15 +225,19 @@ void CApplicationMessenger::ProcessMessage(ThreadMessage *pMsg)
     return;
   }
 
-  CSingleLock lock(m_critSection);
   int mask = pMsg->dwMessage & TMSG_MASK_MESSAGE;
 
-  auto target = m_mapTargets.at(mask);
-  if (target != nullptr)
+  IMessageTarget* target = nullptr;
   {
-    CSingleExit exit(m_critSection);
-    target->OnApplicationMessage(pMsg);
+    CSingleLock lock(m_targetCritSection);
+    auto targetI = m_mapTargets.find(mask);
+    if (targetI != m_mapTargets.end())
+    {
+      target = targetI->second;
+    }
   }
+  if (target != nullptr)
```

Comment:
You did the same mistake again. A copy is worth nothing. At the time you use the pointer, the target might have disappeared already.
(I know that there is no Unregister atm, but this needs fixing)

I bet you can't show any measurable improvement of this change.

Reply:
@FernetMenta the copy is not new, just the way the lock applied. Other guys suggested a loop change and that seemed reasonable.

Reply:
If the lock (used whren register etc,) is not sufficient to protect the list of targets then what is? 
My change is simply to use a separate lock for adding events to the queue, no reason for sending messages to prevent other messages being queued.

Reply:
Dispatching massages here is done by the main/render thread. Nothing must block or hijack this thread for processing intensive tasks. Hence the benefit of having separate locks is nil and I bet that you can't prove the opposite. Even if you could measure any improvement, it would just show that some other code blocks the main thread.
In other words, you code for something that must not happen. In addition you don't fix the obvious errors that is leaving the lock for dispatching.

Reply:
>Hence the benefit of having separate locks is nil and I bet that you can't prove the opposite.

Reply:
@FernetMenta AFAIK I have not added errors (check the original code), but if more can be done then do share what.

Reply:
It must not happen <- if it happens, fix the root cause and don't introduce more complexity via an additional lock.
If something can be done without the lock, please keep using SingleExit.

While your communication is not optimal - the code change is - in my eyes - not an optimization because "optimization of locking" does not mean add more fine grain locking, but REMOVE not needed locking ...

Reply:
Clarify: "your" <- yours and fernet's communication.